### ISSUE TITLE
Marian Encoder's last hidden states from MarianMT and TFMarianMTModel don't match

### ISSUE THREAD

**Issue by harshitadd** [huggingface/transformers#13568]
## Environment info
- `transformers` version: 4.10.2
- Platform: Linux-5.4.104+-x86_64-with-Ubuntu-18.04-bionic
- Python version: 3.7.11
- PyTorch version (GPU?): 1.9.0+cu102 (False)
- Tensorflow version (GPU?): 2.6.0 (False)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

Additional Environment Info: Tensorflow Version - 2.6.0

### Who can help
@patil-suraj, @Rocketknight1
## Information

Model I am using (Bert, XLNet ...): MarianMT and TFMarianMT Models  ( [pretrained opus-mt en-hi model](https://huggingface.co/Helsinki-NLP/opus-mt-en-hi/tree/main)) 

The problem arises when using:
* [ ] my own modified scripts: [Reproducible Error Script](https://colab.research.google.com/drive/1mIEip64duk6upCerF8J1HOoVyN4Dz7sF?usp=sharing)

The tasks I am working on is:
* [ ] my own task or dataset: MT, Not Applicable

## To reproduce

1. Run the provided error script - It can be noticed that the encoder's last hidden state is not the same for MarianMT and TFMarianMT Models. 

## Expected behavior

In my best understanding, The hidden states of the last layer of the encoder should be consistent across both variants ( TFMarianMT and MarianMT ). The variance in the encoder's hidden states will carry forward to introduce difference in the decoded output's generation, potentially leading to different outputs from the same pretrained model.  

Thanks in Advance!

**Comment by Rocketknight1** [comment]
Hi! I ran your script, but I noticed that when I re-ran the code, even with the same framework, I got different outputs. When I set `training=False` for the TF model call and called `model.eval()` for PyTorch this stopped, so I'm guessing it was caused by Dropout layers or something similar.

Even after I did this, though, the output of the two models remained different, so this may indeed be a bug. We're investigating!

**Comment by harshitadd** [comment]
> Hi! I ran your script, but I noticed that when I re-ran the code, even with the same framework, I got different outputs. When I set `training=False` for the TF model call and called `model.eval()` for PyTorch this stopped, so I'm guessing it was caused by Dropout layers or something similar.
> 
> Even after I did this, though, the output of the two models remained different, so this may indeed be a bug. We're investigating!

Thanks for the quick response. I debugged a little myself by looking into generation_tf_utils.py and figured that I got the pytorch equivalent outputs by calling:


encoder = model.get_encoder()
encoded_sequence = encoder(input_ids)

where model is an isntance of the TFMarianMTModel. The key inconsistency then appears to be the difference in this output and the encoder.last_hidden_state. It might also be useful to point this out in the documentation since, prima facie, the attribute does not seem to be returning an expected value.

**Comment by harshitadd** [comment]
Hello @Rocketknight1, Is there any way to get around the final logit bias inconsistency ( in the meantime ) mentioned in #12647 ? I am trying to write the autoregressive loop for generation from scratch and the can't seem to get a consistent output against the logits created by the model forward pass ( which internally add the final bias logits to the TFSharedEmbedding tensor of the decoder last hidden state as evident [here](https://huggingface.co/transformers/_modules/transformers/models/marian/modeling_tf_marian.html#TFMarianMTModel) ) and the logits I create by simply using the last decoder hidden state; [ Minimum Viable Script to reproduce is [here](https://colab.research.google.com/drive/1v5KUSUarg5rB_UOw_v72lbNf4AX_S1qk?usp=sharing))

Thanks a ton!
